File: iosApp/iosApp/Modules/Confirmation ✅/ConfirmationWireframeFactory.swift
// Created by web3d4v on 20/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - ConfirmationWireframeFactory

protocol ConfirmationWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: ConfirmationWireframeContext
    ) -> ConfirmationWireframe
}

// MARK: - DefaultConfirmationWireframeFactory

final class DefaultConfirmationWireframeFactory {
    private let walletService: WalletService
    private let authenticateWireframeFactory: AuthenticateWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let webViewWireframeFactory: WebViewWireframeFactory
    private let deepLinkHandler: DeepLinkHandler
    private let nftsService: NFTsService
    private let mailService: MailService
    private let currencyStoreService: CurrencyStoreService

    init(
        walletService: WalletService,
        authenticateWireframeFactory: AuthenticateWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        webViewWireframeFactory: WebViewWireframeFactory,
        deepLinkHandler: DeepLinkHandler,
        nftsService: NFTsService,
        mailService: MailService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.walletService = walletService
        self.authenticateWireframeFactory = authenticateWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.webViewWireframeFactory = webViewWireframeFactory
        self.deepLinkHandler = deepLinkHandler
        self.nftsService = nftsService
        self.mailService = mailService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultConfirmationWireframeFactory: ConfirmationWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: ConfirmationWireframeContext
    ) -> ConfirmationWireframe {
        DefaultConfirmationWireframe(
            parent,
            context: context,
            walletService: walletService,
            authenticateWireframeFactory: authenticateWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            webViewWireframeFactory: webViewWireframeFactory,
            deepLinkHandler: deepLinkHandler,
            nftsService: nftsService,
            mailService: mailService,
            currencyStoreService: currencyStoreService
        )
    }
}

// MARK: - Assembler

final class ConfirmationWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> ConfirmationWireframeFactory in
            DefaultConfirmationWireframeFactory(
                walletService: resolver.resolve(),
                authenticateWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                webViewWireframeFactory: resolver.resolve(),
                deepLinkHandler: resolver.resolve(),
                nftsService: resolver.resolve(),
                mailService: resolver.resolve(),
                currencyStoreService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/confirmation/ConfirmationWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.confirmation

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframe
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService
import smartadapter.internal.extension.name

interface ConfirmationWireframeFactory {
    fun make(parent: Fragment?, context: ConfirmationWireframeContext): ConfirmationWireframe
}

class DefaultConfirmationWireframeFactory(
    private val walletService: WalletService,
    private val nftsService: NFTsService,
    private val currencyStoreService: CurrencyStoreService,
): ConfirmationWireframeFactory {

    override fun make(
        parent: Fragment?, context: ConfirmationWireframeContext
    ): ConfirmationWireframe = DefaultConfirmationWireframe(
        parent?.let { WeakRef(it) },
        context,
        walletService,
        nftsService,
        currencyStoreService,
    )
}

class ConfirmationWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(ConfirmationWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultConfirmationWireframeFactory(
                it.resolve(WalletService::class.name),
                it.resolve(NFTsService::class.name),
                it.resolve(CurrencyStoreService::class.name),
            )
        }
    }
}